src/main/java/hudson/plugins/ec2/EC2RetentionStrategy.java

```diff
@@ -110,14 +110,23 @@ private long internalCheck(EC2Computer computer) {
             InstanceState state;
 
             try {
+                state = computer.getState(); //Get State before Uptime because getState will refresh the cached EC2 info
                 uptime = computer.getUptime();
-                state  = computer.getState();
             } catch (AmazonClientException | InterruptedException e) {
                 // We'll just retry next time we test for idleness.
                 LOGGER.fine("Exception while checking host uptime for " + computer.getName()
                         + ", will retry next check. Exception: " + e);
                 return 1;
             }
+            
+            //Don't bother checking anything else if the instance is already in the desired state:
+            // * Already Terminated
+            // * We use stop-on-terminate and the instance is currently stopped or stopping
+            if (InstanceState.TERMINATED.equals(state)
+                  || computer.getSlaveTemplate().stopOnTerminate && (InstanceState.STOPPED.equals(state) || InstanceState.STOPPING.equals(state))) {
```

Comment:
I'll defer this to others but I don't think this does the right thing in case the instance was stopped by external forces

Reply:
Can you clarify what you are concerned about?

Reply:
What if someone terminated an instance from AWS console. It now wont call `idletimeout` in those cases

Reply:
Aha, it's true that this will be a slight behavior change in the case of an external EC2 instance state change.

In the case of an external Termination, the difference in behavior is that the node will not be removed from jenkins by this process, but the `EC2SlaveMonitor` thread will do so instead, so that should be OK (perhaps with slightly different timing, but I don't think that should be problematic).

Where there's a slightly more problematic change is the case of an external Stop - there doesn't appear to be an explicit "disconnect" unless you use `hudson.slaves.ConnectionActivityMonitor` (which I forgot isn't enabled by default).
I think it's probably worth a tweak to add an explicit disconnection for this case (and actually this can help us detect disconnections due to external stopping faster).  I will make an update for this.


It seems that